jwst/background/background_sub.py

```diff
@@ -39,53 +49,86 @@ def background_sub(input_model, bkg_list, sigma, maxiters):
 
     # Subtract the average background from the member
     log.debug(' subtracting avg bkg from {}'.format(input_model.meta.filename))
+
     result = subtract_images.subtract(input_model, bkg_model)
 
-    # We're done. Return the result.
+    # We're done. Return the background model and the result.
     return bkg_model, result
 
 
 def average_background(bkg_list, sigma, maxiters):
     """
-    Average multiple background exposures into a combined data model
+    Average multiple background exposures into a combined data model.
+    Processes either 2D (rate) or 3D (rateints) backgrounds.
 
     Parameters:
     -----------
-
     bkg_list: filename list
         List of background exposure file names
 
+    sigma: float, optional
+        Number of standard deviations to use for both the lower
+        and upper clipping limits.
+
+    maxiters: int or None, optional
+        Maximum number of sigma-clipping iterations to perform
+
     Returns:
     --------
-
     avg_bkg: data model
         The averaged background exposure
-
     """
 
+    # Determine the type and dimensionality of the background files
+    try:
+        bkg_try = datamodels.ImageModel(bkg_list[0])
+        image_shape = bkg_try.data.shape
+        bkg_dim = 2
+    except ValueError:
+        try:
+            bkg_try = datamodels.CubeModel(bkg_list[0])
+            image_shape = bkg_try.data.shape[1:]
+            bkg_dim = 3
+            accum_dq_arr = np.zeros((image_shape), dtype=np.int32)
+        except ValueError:
+            pass
+
+    avg_bkg = datamodels.ImageModel(image_shape)
     num_bkg = len(bkg_list)
-    avg_bkg = None
-    cdata = None
+    cdata = np.zeros(((num_bkg,) + image_shape))
+    cerr = cdata.copy()
 
     # Loop over the images to be used as background
     for i, bkg_file in enumerate(bkg_list):
         log.info(f'Accumulate bkg from {bkg_file}')
-        bkg_model = datamodels.ImageModel(bkg_file)
 
-        # Initialize the avg_bkg model, if necessary
-        if avg_bkg is None:
-            avg_bkg = datamodels.ImageModel(bkg_model.shape)
+        if bkg_dim == 2:  # do what the old code does
+            bkg_model = datamodels.ImageModel(bkg_file)  # 2D
```

Comment:
We probably need to change this to a generic `datamodels.open()` rather than forcing it into an `ImageModel`. I didn't realize till now that the old code was doing that. It shouldn't be. Clearly when the step was first written we only envisioned it being used for simple imaging observations. But now we're throwing a number of different model types at it.

Reply:
It's OK to still create `avg_bkg` as an `ImageModel` (up above in line 96), because all we need for the average background data internally is a simple 2D image model.